### PR TITLE
SUNDIALS: fix aliased N_VProd/N_VDiv and the multi-component constraint mask

### LINKED ISSUE #5718: N_VProd/N_VDiv_MultiFab break when z aliases y; N_VConstrMask_MultiFab checks only component 0
**Severity:** high/medium · **Category:** correctness · **Subsystem:** Base-io (blast radius 2/5)

**Locations:** `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:298`, `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:284`, `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:576`

**Based on commit `248fcfb7a5`** (branch `ai_audit`; line numbers refer to that tree).

Three ops in Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp assume conditions the N_Vector contract does not give them: N_VProd and N_VDiv assume the output is not aliased to an input (SUNDIALS aliases z to y in arkUpperBoundH0/cvHin, SPFGMR/SPBCGS/SPTFQMR scaling and constraint step reduction), and N_VConstrMask assumes the mask MultiFab has one component.

## The defect

**`Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:298`** — N_VDiv_MultiFab(x,y,z) is implemented as Copy(z<-x); Divide(z,y), so when SUNDIALS passes z aliased to y (which it does) the result is x/x = 1 (or 0/0 = NaN) instead of x/y.

**`Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:284`** — N_VProd_MultiFab(x,y,z) is Copy(z<-x); Multiply(z,y); with z aliased to y (as SUNDIALS does) it yields x*x instead of x*y.

**`Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:576`** — N_VConstrMask_MultiFab decides 'all constraints satisfied' from mf_m->norm1(), whose default arguments are comp=0, so violations flagged in components >= 1 of a multi-component MultiFab are ignored and SUNTRUE is returned.

## Why it matters

*F038:* Any build with SUNDIALS. SundialsIntegrator::evolve() with integration.use_adaptive_time_step=1 (or a direct CVode/ARKStep run with no user h0) calls arkHin/cvHin -> arkUpperBoundH0, which does N_VDiv(temp2, temp1, temp1) (arkode.c:2694, cvode.c:2314, kinsol.c:2067). temp1 becomes |f|/|f| (1 or NaN) rather than |f|/(0.1|y|+1/ewt), so the initial step-size upper bound hub_inv is silently wrong.

*F190:* Any build with SUNDIALS. AMReX links sunlinsolspfgmr; SUNLinSol_SPFGMR does N_VProd(s1, V[l+1], V[l+1]) (sunlinsol_spfgmr.c:685); SPBCGS/SPTFQMR do N_VProd(sx, x, x); ARKStep/CVODE constraint handling does N_VProd(mm, tmp, tmp) (arkode.c:3406, cvode.c:3306); KINSOL does N_VProd(vtemp2, vtemp1, vtemp1). With a MultiFab N_Vector these produce s1*s1 / mm*mm etc., silently corrupting the Krylov basis or the constraint step reduction.

*F191:* Any build with SUNDIALS. Wrap an nComp>=2 MultiFab as N_Vector, call ARKStepSetConstraints/CVodeSetConstraints with a constraint (e.g. 2.0 = positivity) on component 1 only, and let component 1 go negative in a stage. The kernel writes m=1 at comp 1, but norm1() sums comp 0 (all zeros) -> temp=0 -> returns SUNTRUE; ARKODE accepts the constraint-violating step silently (and arkInitialSetup's y0 check passes).

## How to reach it

- *F038:* config: -DAMReX_SUNDIALS=ON -DSUNDIALS_ROOT=... (any SPACEDIM/precision). Built by .github/workflows/sundials.yml (GCC SUNDIALS 6.5.0; CUDA SUNDIALS 7.0.0, arkode.c:2545 same call) - compile only, no ctest.
- *F038:* API path: ParmParse: integration.type=SUNDIALS, integration.use_adaptive_time_step=1 (AMReX_IntegratorBase.H:172); TimeIntegrator<MultiFab> ti(S); ti.evolve(S, t_out) -> ARKStepEvolve -> arkInitialSetup -> arkHin -> arkUpperBoundH0 -> N_VDiv(temp2,temp1,temp1) -> (ManyVector ->) N_VDiv_MultiFab. Also direct public use: N_VMake_MultiFab(...) + ARKStepCreate/CVodeCreate without SetInitStep.
- *F038:* test coverage: none - Tests/ has no SUNDIALS program; sundials.yml is compile-only. Src caller: Src/Extern/SUNDIALS/AMReX_SundialsIntegrator.H evolve(); nothing asserts on h0. Severity lowered to medium: only the initial-step upper bound hub_inv is wrong (hub falls back to min(0.1*tdist, 1)); NaN fails the `hub*hub_inv > ONE` compare so no NaN enters h, and error control corrects later steps.
- *F190:* config: -DAMReX_SUNDIALS=ON -DSUNDIALS_ROOT=... (any SPACEDIM). Built by .github/workflows/sundials.yml (GCC 6.5.0, CUDA 7.0.0), compile only.
- *F190:* API path: No in-tree caller reaches an aliased N_VProd (integrator uses SPGMR, never sets constraints); public API does: y = N_VMake_MultiFab(len,&mf,&ctx); mem = ARKStepCreate(fe,fi,t0,y,ctx); ARKStepSStolerances(mem,rtol,atol); LS = SUNLinSol_SPFGMR(y,SUN_PREC_NONE,m,ctx); ARKStepSetLinearSolver(mem,LS,NULL); ARKStepEvolve(...) -> spfgmr.c:582. Or ARKStepSetConstraints(mem,c)/CVodeSetConstraints -> arkode.c:2924 / cvode.c:2913 after a violation.
- *F190:* test coverage: none - no Tests/ SUNDIALS program; Src caller AMReX_SundialsIntegrator.H only exercises non-aliased SPGMR/N_VProd paths. No assertion anywhere on the affected values.
- *F191:* config: -DAMReX_SUNDIALS=ON -DSUNDIALS_ROOT=... (any SPACEDIM/precision). Built by .github/workflows/sundials.yml (GCC 6.5.0, CUDA 7.0.0), compile only.
- *F191:* API path: Public API, no in-tree caller (SundialsIntegrator never calls SetConstraints and arkode_mem is private): MultiFab mf(ba,dm,2,0); y = N_VMake_MultiFab(2*ba.numPts(),&mf,&ctx); c = N_VClone(y); N_VConst(0,c); set comp1 of getMFptr(c) to 2.0; mem = ARKStepCreate(...,y,ctx); ARKStepSetConstraints(mem,c); ARKStepEvolve(...) -> N_VConstrMask_MultiFab returns SUNTRUE despite comp-1 negatives. Same via CVodeSetConstraints.
- *F191:* test coverage: none - no Tests/ SUNDIALS program; Src/Extern/SUNDIALS/AMReX_SundialsIntegrator.H does not set constraints. Nothing asserts on the constraint mask result.

## Suggested fix

Replace the Copy-then-Multiply / Copy-then-Divide pairs in N_VProd_MultiFab and N_VDiv_MultiFab with a single elementwise ParallelFor over mfi.validbox() that writes z(i,j,k,c) = x*y (resp. x/y). Reading both inputs before writing the output makes z == y and z == x both correct; this is the same pattern N_VAbs_MultiFab and N_VCompare_MultiFab already use in this file, so no new convention is introduced. Drop the now-unused nghost locals (CI compiles with -Werror). The two hunks are independent; land N_VProd first, then N_VDiv applies at a +10 offset. For N_VConstrMask_MultiFab, sum mf_m->norm1(c, 0) over every component instead of the default norm1(), which reduces comp 0 only. N_VL1Norm_MultiFab in the same file already loops over components this way. Each norm1(c,0) call is a global reduction (local=false), so the trailing ParallelAllReduce::Max becomes redundant; the diff leaves it in place for minimality, and the maintainer may remove it (or switch to a single ReduceOps<ReduceOpMax> over mf_m to avoid ncomp reductions) as a follow-up. Maintainer decisions: whether to add a small SUNDIALS N_Vector unit test under Tests/ (none exists today; sundials.yml is compile-only), which would have caught all three; and whether to audit the remaining ops here for the same alias assumption.

For `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:298` (F038):

```diff
--- a/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
+++ b/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
@@ -292,11 +292,21 @@ void N_VDiv_MultiFab(N_Vector x, N_Vector y, N_Vector z)
     amrex::MultiFab *mf_z = amrex::sundials::getMFptr(z);
 
     int ncomp = mf_x->nComp();
-    //int nghost = mf_x->nGrow();
-    int nghost = 0;  // do not include ghost cells
 
-    amrex::MultiFab::Copy(*mf_z, *mf_x, 0, 0, ncomp, nghost);
-    amrex::MultiFab::Divide(*mf_z, *mf_y, 0, 0, ncomp, nghost);
+    // ghost cells not included; elementwise so that z may alias x or y
+    for (amrex::MFIter mfi(*mf_x); mfi.isValid(); ++mfi)
+    {
+         const amrex::Box& bx = mfi.validbox();
+         amrex::Array4<amrex::Real const> const& x_fab = mf_x->const_array(mfi);
+         amrex::Array4<amrex::Real const> const& y_fab = mf_y->const_array(mfi);
+         amrex::Array4<amrex::Real> const& z_fab = mf_z->array(mfi);
+
+         amrex::ParallelFor(bx, ncomp,
+         [=] AMREX_GPU_DEVICE (int i, int j, int k, int c) noexcept
+         {
+             z_fab(i,j,k,c) = x_fab(i,j,k,c) / y_fab(i,j,k,c);
+         });
+    }
 }
 
 void N_VScale_MultiFab(amrex::Real c, N_Vector x, N_Vector z)
```

<sub>Replaces Copy(z<-x);Divide(z,y) with a single elementwise ParallelFor z=x/y, following the N_VAbs_MultiFab/N_VCompare_MultiFab pattern in this file. Correct for z==y (arkUpperBoundH0/cvHin) and z==x. The unused `nghost` local is removed because CI compiles with -Werror -Wall. Like the twins, no OMP tiling (file convention). Cross-reference: F190 (N_VProd) is the same copy-then-operate aliasing defect in the sibling function; the two diffs stack (F190 notes the +10 offset).</sub>

For `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:284` (F190):

```diff
--- a/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
+++ b/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
@@ -278,11 +278,21 @@ void N_VProd_MultiFab(N_Vector x, N_Vector y, N_Vector z)
     amrex::MultiFab *mf_z = amrex::sundials::getMFptr(z);
 
     int ncomp = mf_x->nComp();
-    //int nghost = mf_x->nGrow();
-    int nghost = 0;  // do not include ghost cells
 
-    amrex::MultiFab::Copy(*mf_z, *mf_x, 0, 0, ncomp, nghost);
-    amrex::MultiFab::Multiply(*mf_z, *mf_y, 0, 0, ncomp, nghost);
+    // ghost cells not included; elementwise so that z may alias x or y
+    for (amrex::MFIter mfi(*mf_x); mfi.isValid(); ++mfi)
+    {
+         const amrex::Box& bx = mfi.validbox();
+         amrex::Array4<amrex::Real const> const& x_fab = mf_x->const_array(mfi);
+         amrex::Array4<amrex::Real const> const& y_fab = mf_y->const_array(mfi);
+         amrex::Array4<amrex::Real> const& z_fab = mf_z->array(mfi);
+
+         amrex::ParallelFor(bx, ncomp,
+         [=] AMREX_GPU_DEVICE (int i, int j, int k, int c) noexcept
+         {
+             z_fab(i,j,k,c) = x_fab(i,j,k,c) * y_fab(i,j,k,c);
+         });
+    }
 }
 
 void N_VDiv_MultiFab(N_Vector x, N_Vector y, N_Vector z)
```

<sub>Same shape as the F038 fix: Copy(z<-x);Multiply(z,y) becomes an elementwise ParallelFor z=x*y, mirroring N_VAbs_MultiFab. Alias-safe for z==y (SPFGMR/SPBCGS/SPTFQMR scaling, constraint step reduction) and z==x. Drops the unused `nghost` local (CI -Werror). Applies independently of F038; stacked, the F038 hunk lands at +10 offset. Cross-reference: F038 (N_VDiv) is the same copy-then-operate aliasing defect in the sibling function; apply both.</sub>

For `Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:576` (F191):

```diff
--- a/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
+++ b/Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp
@@ -573,6 +573,8 @@ int N_VConstrMask_MultiFab(N_Vector a_a, N_Vector a_x, N_Vector a_m)
             });
     }
 
-    temp = mf_m->norm1();
+    for (int c = 0; c < ncomp; ++c) {
+        temp += mf_m->norm1(c, 0);
+    }
     /* Return false if any constraint was violated */
     ParallelAllReduce::Max(temp, ParallelContext::CommunicatorSub());
```

<sub>Sums norm1 over every component instead of comp 0 only, matching N_VL1Norm_MultiFab in the same file; `temp` is already initialized to 0 and `ncomp` is in scope. Each norm1(c,0) is globally reduced (local=false), so the existing ParallelAllReduce::Max stays redundant-but-harmless and is left alone to keep the diff minimal.</sub>

<sub>Diff(s) are against `248fcfb7a5`, written from the current source and verified only with `git apply --check` — never compiled, never run, never applied to the tree. Treat them as precise intent, not tested patches.</sub>

## Verification evidence

<details><summary><code>F038</code> — confirmed (one verifier lens)</summary>

Lens 1 (refutation attempt): Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:298-299 `MultiFab::Copy(*mf_z,*mf_x,...); MultiFab::Divide(*mf_z,*mf_y,...)` with no alias handling (MultiFab::Divide is plain elementwise, AMReX_MultiFab.cpp:279-289). SUNDIALS v6.5.0 (the CI version) arkode.c:2286 `N_VDiv(temp2, temp1, temp1)` in arkUpperBoundH0 and cvode.c:2079 same; reached whenever hin==0 (arkode.c:1904-1925). nvector_manyvector.c:613 forwards subvectors, so the alias reaches N_VDiv_MultiFab. SundialsIntegrator::evolve() (AMReX_SundialsIntegrator.H:815-820) skips SetFixedStep when use_adaptive_time_step and never calls ARKStepSetInitStep, so arkHin runs. Result temp1 = |f|/|f| (1 or NaN) instead of |f|/(0.1|y|+1/ewt). Not fixed on development (git log ai_audit..development -- Src/Extern/SUNDIALS empty); not in KNOWN_STATE.md.

</details>

<details><summary><code>F190</code> — confirmed (one verifier lens)</summary>

Lens 1 (refutation attempt): Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:284-285 `Copy(*mf_z,*mf_x); Multiply(*mf_z,*mf_y)`: with z==y yields x*x. Aliased z==y call sites in libraries AMReX links (Tools/CMake/AMReXThirdPartyLibraries.cmake:181 requires arkode cvode sunlinsolspgmr sunlinsolspfgmr): sunlinsol_spfgmr.c:582 `N_VProd(s1, V[l+1], V[l+1])` (scale1 always true because arkLsSolve sets scaling vectors, arkode_ls.c:2646-2649) -> Krylov vector becomes ewt^2; arkode.c:2924 and cvode.c:2913 `N_VProd(mm, tmp, tmp)` in constraint step reduction. In-tree integrator uses SPGMR whose call is non-aliased (sunlinsol_spgmr.c:597 `N_VProd(s1, vtemp, V[l_plus_1])`) and arkode_mem is private (AMReX_SundialsIntegrator.H:96/139), so in-tree paths are safe; public-API paths are not. Not fixed on development; not in KNOWN_STATE.md.

</details>

<details><summary><code>F191</code> — confirmed (one verifier lens)</summary>

Lens 1 (refutation attempt): Src/Extern/SUNDIALS/AMReX_NVector_MultiFab.cpp:576 `temp = mf_m->norm1();` -> MultiFab::norm1(int comp=0, int ngrow=0, bool local=false) (AMReX_MultiFab.H:249) which reduces only `std::abs(a(i,j,k,comp))` with comp=0 (AMReX_MultiFab.cpp:1165-1191), while the kernel at :555-573 writes m_fab(i,j,k,c)=1 for all c<ncomp. Compare N_VL1Norm_MultiFab (:471-478) which loops norm1(c) over comps. Commit ac4f628498 (#5302) fixed the threshold (==1.0 -> >0.0) but kept norm1(). m is a clone of the state (tempv1, arkode.c:1849/2904) so nComp>=2 whenever the state is; a violation in comp>=1 returns SUNTRUE at arkInitialSetup (arkode.c:1849) and arkCheckConstraints (arkode.c:2904). Not fixed on development; KNOWN_STATE.md lists only #5042 (SUNMemory thread count), a different defect.

</details>

---

<sub>Based on commit `248fcfb7a5`, the tree the audit verified against. From an automated audit of AMReX `Src/`. Audit finding ids: F038, F190, F191. Reviewer unit(s): Base/Integrators-SUNDIALS-SDC-2. Nothing here was compiled or run except where the evidence says so — the failure scenarios are code reasoning, so the reaching configuration above is the cheapest way to confirm or refute it.</sub>

### PR DESCRIPTION
**Pull request by WeiqunZhang**
SUNDIALS calls N_VProd and N_VDiv with the output aliased to an input, but both were Copy(z<-x) followed by Multiply/Divide by y, giving x*x or x/x when z is y. Read both inputs in one elementwise loop instead.

N_VConstrMask used norm1() with its default arguments, so a violation in any component other than 0 was ignored. Sum the local norm of every component; the existing ParallelAllReduce::Max makes it global.

Fixes #5718.

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
# PR #5832: SUNDIALS: fix aliased N_VProd/N_VDiv and the multi-component constraint mask

PR: https://github.com/AMReX-Codes/amrex/pull/5832
Reviewed commit: `9382f99406d5c0c377b78e0139e0c3565444d932`
Reviewed at: 2026-09-07 20:37 UTC
Status: Complete
Result: No actionable findings

## Findings

No actionable findings. Each product/division element reads both inputs before writing the output, preserving exact x/y/z aliasing. The constraint-mask reduction starts at zero, examines every component locally, and then performs the existing global reduction.

## Validation

Inspected PR description, conversation, inline comments, and the complete diff with `git diff ae079342654b60f95effddd4377cd9ffaf573f5f 9382f99406d5c0c377b78e0139e0c3565444d932`. Inspected the complete N_VProd, N_VDiv and N_VConstrMask callbacks and the norm1 argument contract at the reviewed head. The new loops preserve valid-cell-only operations. No local SUNDIALS callback/runtime test was run; z==x, z==y, all-equal vectors, later-component constraints and GPU execution remain runtime coverage limits.

CI at reviewed head: SUCCESS: 75. CI status is separate from the source assessment.

Reviewed by gpt-6-astra.

**Review by asalmgren — APPROVED** [review]
